MetaLamp/lending-pool/src/Plutus/Contracts/Endpoints.hs

```diff
@@ -0,0 +1,225 @@
+
+valueAt :: HasBlockchainActions s => Address -> Contract w s Text Value
+valueAt address = do
+    os <- map snd . Map.toList <$> utxoAt address
+    pure $ mconcat [txOutValue $ txOutTxOut o | o <- os]
+
+fundsAt :: HasBlockchainActions s => PubKeyHash -> Contract w s Text Value
+fundsAt pkh = valueAt (pubKeyHashAddress pkh)
+
+balanceAt :: HasBlockchainActions s => PubKeyHash -> AssetClass -> Contract w s Text Integer
+balanceAt pkh asset = flip assetClassValueOf asset <$> fundsAt pkh
+
+poolFunds :: HasBlockchainActions s => Aave -> Contract w s Text Value
+poolFunds aave = valueAt (Core.aaveAddress aave)
+
+data DepositParams =
+  DepositParams {
+    dpAsset      :: AssetClass,
+    dpOnBehalfOf :: PubKeyHash,
+    dpAmount     :: Integer
+  }
+    deriving stock    (Show, Generic)
+    deriving anyclass (ToJSON, FromJSON, ToSchema)
+
+PlutusTx.unstableMakeIsData ''DepositParams
+PlutusTx.makeLift ''DepositParams
+
+deposit :: (HasBlockchainActions s) => Aave -> DepositParams -> Contract w s Text ()
+deposit aave DepositParams {..} = do
+    reserveOutput <- State.findAaveReserve aave dpAsset
+    let reserve = soDatum reserveOutput
+        lookups = Constraints.ownPubKeyHash dpOnBehalfOf
+            <> Constraints.scriptInstanceLookups (Core.aaveInstance aave)
+        outValue = assetClassValue (rCurrency reserve) dpAmount
+        tx = mustPayToTheScript Core.DepositDatum outValue
+    ledgerTx <- submitTxConstraintsWith lookups tx
+    _ <- awaitTxConfirmed $ txId ledgerTx
+
+    wasZeroBalance <- (== 0) <$> balanceAt dpOnBehalfOf (rAToken reserve)
+    _ <- AToken.forgeATokensFrom aave reserve dpOnBehalfOf dpAmount
+    when wasZeroBalance $ do
+        userOutputs <- State.findOutputsBy aave (State.userStateToken aave) State.pickUserConfig
+        case userOutputs of
+            [] -> void $
+                State.putUser aave $ UserConfig { ucAddress = dpOnBehalfOf, ucReserveId = rCurrency reserve, ucUsingAsCollateral = True }
+            [userOutput] -> void $
+                State.updateUser aave $ Prelude.fmap (\u -> u { ucUsingAsCollateral = True }) userOutput
+            _ -> throwError "Invalid state: multiple users"
+
+    void $ State.updateReserve aave $ Prelude.fmap (\r -> r { rAmount = rAmount r + dpAmount }) reserveOutput
+
+data WithdrawParams =
+    WithdrawParams {
+        wpAsset  :: AssetClass,
+        wpTo     :: PubKeyHash,
+        wpFrom   :: PubKeyHash,
+        wpAmount :: Integer
+    }
+    deriving stock    (Show, Generic)
+    deriving anyclass (ToJSON, FromJSON, ToSchema)
+
+PlutusTx.unstableMakeIsData ''WithdrawParams
+PlutusTx.makeLift ''WithdrawParams
+
+withdraw :: (HasBlockchainActions s) => Aave -> WithdrawParams -> Contract w s Text ()
+withdraw aave WithdrawParams {..} = do
+    reserveOutput <- State.findAaveReserve aave wpAsset
+    let reserve = soDatum reserveOutput
+
+    balance <- balanceAt wpFrom (rAToken reserve)
+    when (wpAmount == balance) $ do
```

Comment:
Seems like `wpAmount <= balance` would be more logical, or do we always burn all aTokens?